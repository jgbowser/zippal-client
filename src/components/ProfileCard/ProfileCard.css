.ProfileCard_main {
  border: 1px solid bisque;
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 75vw;
  margin: 0 auto;
  padding: 12px;
  background-image: linear-gradient(220deg, white, darkkhaki);
  box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19);
}

.ProfileCard__icon_container {
  width: 50%;
  margin: auto 0;
}

.ProfileCard__bio_location {
  text-align: left;
  margin: auto 10px;
}

.ProfileCard__label {
  line-height: 2em;
  text-align: left;
  font-size: 18px;
  color: #3a392a;
}

.ProfileCard__data {
  margin-left: 10px;
  font-size: 20px;
}
