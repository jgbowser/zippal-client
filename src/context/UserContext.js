import React, {Component} from 'react'
import AuthApiService from '../services/auth-api-service'
import UserService from '../services/user-api-service'
import TokenService from '../services/token-service'
import IdleService from '../services/idle-service'

const UserContext = React.createContext({
  user:{},
  profileInfo:{},
  error:null,
  setError:() => { },
  clearError: () => { },
  setUser: () => { },
  getProfile: () => { },
  updateProfile: () => { },
  processLogin: () => { },
  processLogout: () => { },
})

export default UserContext

export class UserProvider extends Component {
  constructor(props){
    super(props)
    const state = {user: {}, error:null, profileInfo: {}}

    const jwtPayload = TokenService.parseAuthToken()
    

    if(jwtPayload){
      const profileData = this.getProfile()

      state.user = {
        id: jwtPayload.id,
        display_name: jwtPayload.display_name,
        username: jwtPayload.sub,
      }

      if(profileData)
        state.profileInfo = {
          bio: profileData.bio,
          display_name: profileData.display_name,
          fa_icon: profileData.fa_icon,
          location: profileData.location
      }
    }
      

    this.state = state;
    IdleService.setIdleCallback(this.logoutBecauseIdle)
  }

  componentDidMount(){
    if(TokenService.hasAuthToken()){
      IdleService.registerIdleTimerResets()
      TokenService.queueCallbackBeforeExpiry(() => {
        this.fetchRefreshToken()
      })
    }
  }

  componentWillUnmount() {
    IdleService.unRegisterIdleResets()
    TokenService.clearCallBackBeforeExpiry()
  }

  setError = error => {
    console.error(error)
    this.setState({error})
  }

  clearError = () => {
    this.setState({error: null})
  }
  
  setUser = user => {
    this.setState({user})
  }

  getProfile = async () => {
    const profileInfo = await UserService.getUserProfile()
    this.setState({profileInfo})
  }

  updateProfile = (profileInfo) => {
    this.setState({profileInfo})
  }

  processLogin = authToken => {
    TokenService.saveAuthToken(authToken)

    const jwtPayload = TokenService.parseAuthToken()
    this.setUser({
      id: jwtPayload.id,
      display_name: jwtPayload.display_name,
      username: jwtPayload.sub,
    })
    this.getProfile()
    IdleService.registerIdleTimerResets()
    TokenService.queueCallbackBeforeExpiry(() => {
      this.fetchRefreshToken()
    })
  }

  processLogout = () => {
    TokenService.clearAuthToken()
    TokenService.clearCallBackBeforeExpiry()
    IdleService.unRegisterIdleResets()
    this.setUser({})
    this.setState({profileInfo: {}})
  }

  logoutBecauseIdle = () => {
    TokenService.clearAuthToken()
    TokenService.clearCallBackBeforeExpiry()
    IdleService.unRegisterIdleResets()
    this.setUser({idle:true})
  }

  fetchRefreshToken = () => {
    AuthApiService.refreshToken()
      .then(res => {
        TokenService.saveAuthToken(res.authToken)
        TokenService.queueCallbackBeforeExpiry(() => {
          this.fetchRefreshToken()
        })
      })
      .catch(err => {
        this.setError(err)
      })
  }

  render() {
    const value = {
      user: this.state.user,
      profileInfo: this.state.profileInfo,
      error: this.state.error,
      setError: this.setError,
      clearError: this.clearError,
      setUser: this.setUser,
      getProfile: this.getProfile,
      updateProfile: this.updateProfile,
      processLogin: this.processLogin,
      processLogout: this.processLogout
    }
    return (
      <UserContext.Provider value={value}>
        {this.props.children}
      </UserContext.Provider>
    )
  }
}