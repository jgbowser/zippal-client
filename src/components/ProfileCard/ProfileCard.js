import React, {Component} from 'react'
import UserContext from '../../context/UserContext'
import UserService from '../../services/user-api-service'
import {library} from '@fortawesome/fontawesome-svg-core'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'

import './ProfileCard.css'


export default class ProfileCard extends Component {


  static contextType = UserContext


  render() {
    const {fa_icon, username, display_name, location, bio} = this.context.profileInfo
    const defaultIcon = 'user-circle'
    return (
        <section className="ProfileCard">
          <h2>Welcome {display_name}!</h2>
          <div className="ProfileCard_main">
            <div className="ProfileCard__icon_container">
              <FontAwesomeIcon className='ProfileCard__icon_fa fa-8x' icon={fa_icon ? fa_icon : defaultIcon} />
              <p>{username}</p>
            </div>
            
            <div className="ProfileCard__bio_location">
              <p>
                <span className="ProfileCard__label">location:</span> <br/> 
                <span className="ProfileCard__data">{location ? location : 'none'}</span>
              </p>
              <p><span className="ProfileCard__label">bio:</span> <br/> 
              <span className="ProfileCard__data">{bio ? bio : 'not available'}</span></p>
            </div>
            {this.renderEditButton()}
          </div>
        </section>
      
    )
  }
}