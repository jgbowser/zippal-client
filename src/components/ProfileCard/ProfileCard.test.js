import React from 'react'
import ReactDOM from 'react-dom'
import UserContext from '../../context/UserContext'
import ProfileCard from './ProfileCard'
import renderer from 'react-test-renderer'

describe('ProfileCard Component', () => {
  it('renders without crashing', () => {
    const profileContext = {
      fa_icon: 'user-circle',
      username: 'test_user',
      display_name: 'Bob',
      location: 'USA',
      bio: 'Test bio'
    }
    const div = document.createElement('div')
    ReactDOM.render(
      <UserContext.Provider value={{profileInfo: profileContext}}>
        <ProfileCard />
      </UserContext.Provider>, div
    )
    ReactDOM.unmountComponentAtNode(div)
  })
  it(`takes a snapshot of the ProfileCard component`, () => {
    const profileContext = {
      fa_icon: 'user-circle',
      username: 'test_user',
      display_name: 'Bob',
      location: 'USA',
      bio: 'Test bio'
    }
    const tree = renderer
    .create(<UserContext.Provider value={{profileInfo: profileContext}}><ProfileCard/></UserContext.Provider>)
    expect(tree).toMatchSnapshot()
  })
})

//{fa_icon, username, display_name, location, bio} = this.context.profileInfo