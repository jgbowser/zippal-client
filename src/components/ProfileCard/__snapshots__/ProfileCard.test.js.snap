// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileCard Component takes a snapshot of the ProfileCard component 1`] = `
<section
  className="ProfileCard"
>
  <h2>
    Welcome 
    Bob
    !
  </h2>
  <div
    className="ProfileCard_main"
  >
    <div
      className="ProfileCard__icon_container"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-user-circle fa-w-16 ProfileCard__icon_fa fa-8x"
        data-icon="user-circle"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 496 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M248 8C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zm0 96c48.6 0 88 39.4 88 88s-39.4 88-88 88-88-39.4-88-88 39.4-88 88-88zm0 344c-58.7 0-111.3-26.6-146.5-68.2 18.8-35.4 55.6-59.8 98.5-59.8 2.4 0 4.8.4 7.1 1.1 13 4.2 26.6 6.9 40.9 6.9 14.3 0 28-2.7 40.9-6.9 2.3-.7 4.7-1.1 7.1-1.1 42.9 0 79.7 24.4 98.5 59.8C359.3 421.4 306.7 448 248 448z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
      <p>
        test_user
      </p>
    </div>
    <div
      className="ProfileCard__bio_location"
    >
      <p
        className="ProfileCard__label"
      >
        location:
      </p>
      <p
        className="ProfileCard__data"
      >
        USA
      </p>
      <p
        className="ProfileCard__label"
      >
        bio:
      </p>
      <p
        className="ProfileCard__data"
      >
        Test bio
      </p>
    </div>
    <button
      className="form_button"
      onClick={[Function]}
      type="button"
    >
      Edit your Profile
    </button>
  </div>
</section>
`;
